include/net/can.h

```diff
@@ -0,0 +1,215 @@
+#define NET_CAN_DAD_ADDR            0x3DFE
+#define NET_CAN_ETH_TRANSLATOR_ADDR 0x3DF0
+#define NET_CAN_MAX_ADDR            0x3DEF
+#define NET_CAN_MIN_ADDR            0x0100
+
+#define CAN_NET_IF_ADDR_MASK       0x3FFF
+#define CAN_NET_IF_ADDR_BYTE_LEN   2U
+#define CAN_NET_IF_ADDR_DEST_POS   14U
+#define CAN_NET_IF_ADDR_DEST_MASK  (CAN_NET_IF_ADDR_MASK << CAN_NET_IF_ADDR_DEST_POS)
+#define CAN_NET_IF_ADDR_SRC_POS    0U
+#define CAN_NET_IF_ADDR_SRC_MASK   (CAN_NET_IF_ADDR_MASK << CAN_NET_IF_ADDR_SRC_POS)
+#define CAN_NET_IF_ADDR_MCAST_POS  28U
+#define CAN_NET_IF_ADDR_MCAST_MASK (1UL << CAN_NET_IF_ADDR_MCAST_POS)
+
+#define CAN_NET_IF_IS_MCAST_BIT    (1U << 14)
+
+#define CAN_NET_FILTER_NOT_SET -1
+
+/** @endcond */
+
+/*
+ *             +-----------+           +-----------+
+ *             |           |           |           |
+ *             |   IPv6    |           |   IPv6    |
+ *             |  6LoCAN   |           |  6LoCAN   |
+ *             |           |           |           |
+ *             +----+-+----+           +----+-+----+
+ *                  | |                     | |                   +---+
+ *  +----+          | |                     | |                  /     \   +-+
+ *  |    |          | |                     | |          +--+   /       \_/   \
+ * +++   +---+   +--+----+   +-------+   +--+----+   +--/    \_/               \
+ * | |        \ /     |   \ /         \ /     |   \ /  /                       |
+ * | |         X      |    X    CAN    X      |    X  |         Internet       |
+ * | |        / \     |   / \         / \     |   / \  \                       /
+ * +++   +---+   +----+--+   +-------+   +----+--+   +--+                     /
+ *  |    |                                               +-------------------+
+ *  +----+
+ */
+struct net_can_api {
+	/**
+	 * The net_if_api must be placed in first position in this
+	 * struct so that we are compatible with network interface API.
+	 */
+	struct net_if_api iface_api;
+
+	/** Send a single CAN frame */
+	int (*send)(struct device *dev, const struct zcan_frame *frame,
+		    can_tx_callback_t cb, void *cb_arg, s32_t timeout);
+	/** Attach a filter with it's callback */
+	int (*attach_filter)(struct device *dev, can_rx_callback_t cb,
+			     void *cb_arg, const struct zcan_filter *filter);
+	/** Detach a filter */
+	void (*detach_filter)(struct device *dev, int filter_id);
+	/** Enable or disable the reception of frames for net CAN */
+	int (*enable)(struct device *dev, bool enable);
+};
+
+/** @cond INTERNAL_HIDDEN */
+
+#define CANBUS_L2_CTX_TYPE	struct net_can_context *
+
+/**
+ * Context for canbus net device.
+ */
+struct canbus_net_ctx {
+	/** Filter ID for link layer duplicate address detection. */
+	int dad_filter_id;
+	/** Work item for responding to link layer DAD requests. */
+	struct k_work dad_work;
+	/** The interface associated with this device */
+	struct net_if *iface;
+	/** The link layer address chosen for this interface */
+	u16_t ll_addr;
```

Comment:
Because Zephyr expects link-layer address as a u8_t pointer. The upper two bits are always zero. This address is the address when you call net_if_get_link_addr().